Rework migrate data docs (Take 2) (#5244)
elastic/docs-content: manage-data/migrate/migrate-data-using-reindex-api.md

For more advanced use cases, including data modification using scripts or ingest pipelines, refer to the [Reindex indices examples](elasticsearch://reference/elasticsearch/rest-apis/reindex-indices.md#reindex-from-remote) and the [reindex API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-reindex) documentation. 

:::{tip}
If you're migrating from a self-managed cluster that uses non–publicly trusted TLS certificates, including self-signed certificates and certificates signed by a private certificate authority (CA), refer to our guide [Reindex from a self-managed cluster using a private CA](/manage-data/migrate/migrate-from-a-self-managed-cluster-with-a-self-signed-certificate-using-remote-reindex.md).
:::

## Prerequisites [migrate-reindex-from-remote-prereqs]

- An {{ech}} deployment with data to migrate
- A [{{serverless-full}}](/deploy-manage/deploy/elastic-cloud/serverless.md) project configured and running
- An [API key](/deploy-manage/api-keys/elastic-cloud-api-keys.md) for authentication with the {{ech}} deployment

  Basic authentication can be used in place of an API key, but an API key is recommended as a more secure option.

> - The target deployment must be able to access your original source cluster to perform the reindex operation. Access is controlled by the {{es}} `reindex.remote.whitelist` user setting.

Reviewer: This section does not apply for serverless destinations - in serverless, we automatically allow access to all ECH endpoints as source. But this is great information and it needs to be kept somewhere. Do you think we could take this (and perhaps the tip about the certificates above) and shift it to a different section of this page? Somewhere clearly delineated from the guided example?

Author: For sure! I've put this in a new "Notes for migrating between other deployment types" section at the bottom of the page, along with the note for folks using private CA.

    2. In the **Elasticsearch** section, select **Manage user settings and extensions**. For deployments with existing user settings, you might have to expand the **Edit elasticsearch.yml** caret for each node type instead.
    3. Add the following `reindex.remote.whitelist: [REMOTE_HOST:PORT]` user setting, where `REMOTE_HOST` is a pattern matching the URL for the remote {{es}} host that you are reindexing from, and `PORT` is the host port number. Do not include the `https://` prefix.

        If you override the parameter, it replaces the defaults: `["*.io:*", "*.com:*"]`. If you still want these patterns to be allowed, you need to specify them explicitly in the value.

        For example:

        `reindex.remote.whitelist: ["*.us-east-1.aws.found.io:9243", "*.com:*"]`

    4. Save your changes.

:::{important} 
Kibana assets must be migrated separately using the {{kib}} [export/import APIs](https://www.elastic.co/docs/api/doc/kibana/group/endpoint-saved-objects) or recreated manually. Refer to [Migration options](/manage-data/migrate-data.md#migration-options) for details about migrating different types of {{es}} data.

> Index templates, data stream definitions, and ILM policies, must be in place _before_ you start reindexing data. However, be mindful that if you have any [ingest pipelines](/manage-data/ingest/transform-enrich/ingest-pipelines.md) configured, it's typically best to add these _after_ data migration so as to avoid re-transforming data that had already been transformed at the time that it was ingested into your source deployment (though if the data is idempotent, re-transforming is not a concern).

Reviewer: ILM does not apply for serverless - perhaps we should mention DLM too? Or just generalize with something like "data lifecycle settings" ?

---

> navigation_title: Migrate data using the reindex API

Reviewer 2: consider updating the link over here to this page instead

https://docs-v3-preview.elastic.dev/elastic/docs-content/pull/5244/deploy-manage/deploy/elastic-cloud/differences-from-other-elasticsearch-offerings#elasticsearch

applies_to:
  serverless: preview

>   stack: all

Reviewer 2: pref `ga` over `all` (this gives a wiggly line if you're using the [docs utilities](https://www.elastic.co/docs/contribute-docs/vscode-extension))

Visual components, such dashboard and visualizations, can be migrated after you have migrated the data.
:::

## Migrate documents from {{ech}} to {{serverless-short}} [migrate-reindex-from-remote-ech-serverless]

1. In your {{ech}} deployment:

    1. Navigate to the deployment home page and copy the {{es}} endpoint. 

    1. Go to the **Index Management** page in the navigation menu or use the [global search field](/explore-analyze/find-and-organize/find-apps-and-objects.md).

    1. Use the search field to identify the indices that you want to migrate.

1. In your {{serverless-short}} project:

    1. Open the Developer Tools [Console](/explore-analyze/query-filter/tools/console.md).

>     1. Call the reindex API to migrate your index. If you have multiiple indices to migrate, we recommend doing a separate call for each.

Reviewer 2: ```suggestion
    1. Call the reindex API to migrate your index. If you have multiple indices to migrate, then you should perform a separate call for each.
```

        1. The API key for authenticating the connection to your {{ech}} deployment.
        1. The source index to copy from your {{ech}} deployment.
        1. The destination index in your {{serverless-short}} project.

    1. Verify that the new index is present:

        ```sh
        GET INDEX-NAME/_search?pretty
        ```

    1. If you are not planning to reindex more data from the remote and you configured a `reindex.remote.whitelist` user setting, that setting can now be removed.

## Notes for migrating between other deployment types [migrate-reindex-from-remote-others]

The page demonstrates copying data from an {{ech}} deployment to {{serverless-short}}. When you use the reindex API to copy data across other deployment types, consider the following:

  - If you're migrating from a self-managed cluster that uses non–publicly trusted TLS certificates, including self-signed certificates and certificates signed by a private certificate authority (CA), refer to our guide [Reindex from a self-managed cluster using a private CA](/manage-data/migrate/migrate-from-a-self-managed-cluster-with-a-self-signed-certificate-using-remote-reindex.md).

>   - The target deployment must be able to access your original source cluster to perform the reindex operation. When you migrate to {{serverless-short}}, access to all {{ech}} endpoints is allowed automatically. For migrating to other deployment types, access is controlled by the {{es}} `reindex.remote.whitelist` user setting.

Reviewer 2: there are only two bullets here. consider subheadings instead

>   - The target deployment must be able to access your original source cluster to perform the reindex operation. When you migrate to {{serverless-short}}, access to all {{ech}} endpoints is allowed automatically. For migrating to other deployment types, access is controlled by the {{es}} `reindex.remote.whitelist` user setting.

Author: I've opened https://github.com/elastic/docs-content-internal/issues/1124 to "generalize" the two tutorials.

:::

For more advanced use cases, including data modification using scripts or ingest pipelines, refer to the [Reindex indices examples](elasticsearch://reference/elasticsearch/rest-apis/reindex-indices.md#reindex-from-remote) and the [reindex API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-reindex) documentation.

## Prerequisites [migrate-reindex-from-remote-prereqs]

- An {{ech}} deployment with data to migrate
- An [{{serverless-full}}](/deploy-manage/deploy/elastic-cloud/serverless.md) project configured and running
- An [API key](/deploy-manage/api-keys/elastic-cloud-api-keys.md) for authentication with the {{ech}} deployment

  Basic authentication can be used in place of an API key, but an API key is recommended as a more secure option.

> - The target deployment must be able to access your original source cluster to perform the reindex operation. Access is controlled by the {{es}} `reindex.remote.whitelist` user setting.

Reviewer 2: As an add-on to pete's comment, maybe we need to either remove this prereq from this list, or we need to preface it with "For all source clusters other than those on ECH" or something. with the current page structure, I'd just take it out, because we specify ECH and serverless instances in the prereqs list.

Author: Good idea. I've removed it.

        Example: Reindex from an {{ech}} deployment to a {{serverless-short}} project using an API key:

        ```
        POST _reindex
        {
          "source": {
            "remote": {
              "host": "https://<SERVERLESS_HOST_URL>:443", <1>
              "api_key": "<ECH_API_KEY>" <2>
            },
            "index": "<SOURCE_INDEX>" <3>
          },
          "dest": {
            "index": "<DESTINATION_INDEX>" <4>
          }
        }
        ```
        1. The URL for your {{serverless-short}} project. This is the {{es}} endpoint that you copied in Step 1. If you're migrating to, for example, an {{ech}} cluster, you can modify the remote host address accordingly.
        1. The API key for authenticating the connection to your {{ech}} deployment.
        1. The source index to copy from your {{ech}} deployment.
        1. The destination index in your {{serverless-short}} project.

    1. Verify that the new index is present:

        ```sh
        GET INDEX-NAME/_search?pretty
        ```

    1. If you are not planning to reindex more data from the remote and you configured a `reindex.remote.whitelist` user setting, that setting can now be removed.

> ## Notes for migrating between other deployment types [migrate-reindex-from-remote-others]

Reviewer 2: 🤔 would this be needed with serverless as the source, or ech with an ech target?

Author: Well, I think each individual note should start with the migration scenario that it applies to ("If you're migrating from X to Y..."). There likely will be other notes to add over time, so personally I think we can leave this section as is.
